### ISSUE TITLE
ftrace -s ./programm            doesnt work with gcc

### ISSUE THREAD

**Issue by flaskymasky** [elfmaster/ftrace#5]
i tried it with gcc on both 32 and 64 executable.
it works with clang tho.

Ubuntu kernel: 5.8.0-59

**Comment by elfmaster** [comment]
Keep in mind it does not work on position independent executables. Back
when I designed it there were virtually none. Also it's very slow it should
not use PTRACE_SINGLESTEP, it should set breakpoints on the function
addresses and use PTRACE_CONT until it hits each breakpoint. Feel free to
update it :)

-Ryan

**Comment by elfmaster** [comment]
What do you mean it doesn't work with gcc? Show me the output.

-Ryan

**Comment by flaskymasky** [comment]
Sorry i had no internet the rest of my text didn't send;

if I run 64 bit

 gcc -o prog prog.c 
ftrace -s prog

output just :

[+] Function tracing begins here:
- nothing -

just the normal program output nothing else.

when I use clang to compile it works perfect, maybe some difference in linking  and symbol tables, I don't know.
But it works otherwise perfectly.
Anyways very good tool and great book you got deep knowledge.

**Comment by flaskymasky** [comment]
Hey, can you recommend me some books for expanding my knowledge in general ( binary analysis, Assembly, ... ) maybe some of which you gained your knowledge too.
I would appreciate it very much.
Thanks

**Comment by elfmaster** [comment]
Hey Flaskymasky,

I mainly learned by studying papers. I wrote a book that has helped alot of
people to learn ELF binary voodoo it's called "learning linux binary
analysis"


-Ryan

**Comment by flaskymasky** [comment]
Ok thanks.
I am doing your book right now, as I said great book and thank you for your time :)

**Comment by elfmaster** [comment]
Absolutely. Best of luck on your journey. Feel free to hit me up with
questions anytime. I'm not sure why gcc is not working to build the
software properly.
One day I may re-write ftrace :)

-Ryan

**Comment by flaskymasky** [comment]
Thank you alot.

**Comment by yhnu** [comment]
@elfmaster 
the problem is that :
![image](https://user-images.githubusercontent.com/23090916/131942074-20036f24-4967-4af1-b43e-f6a8be42b833.png)

and some call instruinstruction is ff15 not prefix is e8

![image](https://user-images.githubusercontent.com/23090916/131942214-17111dfe-1915-4eac-adbc-78f3c94f54f5.png)

**Comment by elfmaster** [comment]
Yes I am familiar w this limitation. It doesn’t handle indirect calls, or
position independent executables.

**Comment by elfmaster** [comment]
If you want to build support I’ll answer any questions you have

**Comment by huangxianxia** [comment]
> @elfmaster the problem is that : ![image](https://user-images.githubusercontent.com/23090916/131942074-20036f24-4967-4af1-b43e-f6a8be42b833.png)
> 
> and some call instruinstruction is ff15 not prefix is e8
> 
> ![image](https://user-images.githubusercontent.com/23090916/131942214-17111dfe-1915-4eac-adbc-78f3c94f54f5.png)

how solve it? thanks,i have same problem

**Comment by huangxianxia** [comment]
> > @elfmaster the problem is that : ![image](https://user-images.githubusercontent.com/23090916/131942074-20036f24-4967-4af1-b43e-f6a8be42b833.png)
> > and some call instruinstruction is ff15 not prefix is e8
> > ![image](https://user-images.githubusercontent.com/23090916/131942214-17111dfe-1915-4eac-adbc-78f3c94f54f5.png)
> 
> how solve it? thanks,i have same problem

answer myself:
gcc option: gcc -no-pie -o test test.c.
![image](https://user-images.githubusercontent.com/2484499/155295121-ba7949a5-8f1b-4cf4-8aec-83be9b9a800a.png)
